### PR TITLE
server,cachewarm: add a mechanism to fetch sql leases upon startup

### PR DESCRIPTION
**Pull request by ajwerner**
Pre-fetching the leases is important to reduce the overall startup latency. The code, as it stands, is hacky in that it does not have a concurrency limit or a limit on the total number of descriptors it'll fetch. It's off by default, so that's okay.

Epic: CRDB-18596

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/cachewarm/leases.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		entries, err = collection.Direct().ScanAllNamespaceEntries(ctx, txn)
+		return err
+	}); err != nil {
+		if ctx.Err() == nil {
+			log.Warningf(ctx, "failed to read catalog")
+		}
+		return
+	}
+
+	log.Infof(ctx, "pre-fetching descriptor leases")
```
Comment (postamar) [inline pkg/sql/catalog/cachewarm/leases.go:76]: nit: consider logging the number of descriptor leases we're going to pre-fetch

File: pkg/sql/catalog/cachewarm/leases.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	_ = entries.ForEachNamespaceEntry(func(e nstree.NamespaceEntry) error {
+		return stopper.RunAsyncTask(ctx, "pre-fetch descriptor", func(ctx context.Context) {
+			desc, _ := lm.Acquire(ctx, db.Clock().Now(), e.GetID())
+			if desc != nil {
+				desc.Release(ctx)
+			}
+		})
+	})
```
Comment (postamar) [inline pkg/sql/catalog/cachewarm/leases.go:84]: I know the point of goroutines is that they're cheap etc etc but I'm uneasy about this being unbounded. Can you set a constant cap in this PR? The ranking you mention in your TODO seems simple enough to implement: sort the namespace entries by:
1. parentID == 1 or not
2. parentID == 0 && parentSchemaID == 0 or not
3. by decreasing ID or whatever, I don't know;

and then fire up goroutines only for the first 1000 or so and ignore the rest.

File: pkg/sql/catalog/internal/catkv/catalog_reader.go [hunk 3]
```diff
@@ -140,6 +143,24 @@ func (cr catalogReader) ScanAll(ctx context.Context, txn *kv.Txn) (nstree.Catalo
 	return mc.Catalog, nil
 }
 
+// ScanAllNamespaceEntries is part of the CatalogReader interface.
+func (cr catalogReader) ScanAllNamespaceEntries(
+	ctx context.Context, txn *kv.Txn,
+) (nstree.Catalog, error) {
+	var mc nstree.MutableCatalog
+	log.Eventf(ctx, "fetching all descriptors and namespace entries")
```
Comment (postamar) [inline pkg/sql/catalog/internal/catkv/catalog_reader.go:151]: s/fetching all descriptors and namespace entries/fetching all namespace entries/

File: pkg/sql/catalog/cachewarm/leases.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cachewarm
```
Comment (postamar) [inline pkg/sql/catalog/cachewarm/leases.go:11]: Perhaps `cachewarm` should be a child package of `lease`? No strong feelings.

**Review by postamar — CHANGES_REQUESTED** [review]

**Comment by ajwerner** [comment]
This is no longer needed.

**Pull request closed without merging**